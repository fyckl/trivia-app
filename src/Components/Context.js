import React, { useContext, useState, useEffect } from 'react'

const Context = React.createContext()

function ContextProvider(props) {

    const [startQuiz, setStartQuiz] = useState(false)
    const [arrQuestions, setArrQuestions] = useState([])
    const [playAgain, setPlayAgain] = useState(false)
    const [correctAnswers, setCorrectAnswers] = useState(0)
    const [score, setScore] = useState(0)
    const [shuffleAnswers, setShuffleAnswers] = useState(false)

    function quizStart(){
      setShuffleAnswers(true)
      setStartQuiz(true)
    }

    function checkAnswers(){
      setShuffleAnswers(false)
      setPlayAgain(true)
    }

    useEffect(() => {
      setScore(prevState => prevState + correctAnswers)
    }, [correctAnswers])

    function reset(){
      setArrQuestions([])
      getQuestions()
      setPlayAgain(false)
      setCorrectAnswers(0)
      setShuffleAnswers(true)
    }

    function getQuestions(){
      fetch('https://opentdb.com/api.php?amount=5&difficulty=easy&type=multiple&encode=base64')
      .then(res=> res.json())
      .then(data=> setArrQuestions(data.results))
    }

    
    // console.log(arrQuestions)
    useEffect(() => {
      getQuestions()
    }, [])

    return(
        <Context.Provider 
            value={{
                quizStart, 
                startQuiz, 
                arrQuestions,
                setCorrectAnswers,
                checkAnswers,
                playAgain,
                correctAnswers,
                reset,
                score,
                shuffleAnswers
        }}>
            {props.children}
        </Context.Provider>
    )
}

export {ContextProvider, Context}