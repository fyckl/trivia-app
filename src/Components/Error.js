import React from 'react'

export default function Error() {
  return (
    <h3 className='karla text-center'>Internal server error, please try again later (┬┬﹏┬┬)</h3>
  )
}
