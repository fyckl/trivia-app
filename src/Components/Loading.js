import React from 'react'
import gif from '../Media/loading.gif'

export default function Loading() {
  return (
    <img src={gif} className="loading flex" alt="loading-gif" />
  )
}
