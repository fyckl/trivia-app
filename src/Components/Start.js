import React from 'react'

export default function Start(props) {
    
    return (
        <div className='text-center questions-container' data-testid="container">
                <h2 className='karla' data-testid="header">Quizzical</h2>
                <h3 className='karla' data-testid="header-2">A game of trivia</h3>
                <button 
                    name='start-quiz' 
                    className='start-button' 
                    onClick={props.start}
                    data-testid="button"
                >
                    Start Quiz
                </button>
        </div>
    )
}
