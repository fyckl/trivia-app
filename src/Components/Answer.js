import React, { useContext } from 'react'
import { Context, useAppContext } from "./Context"
import { Buffer } from 'buffer'

export default function Answer({ answers, correctAnswer}) {

  const {
    isPlayAgain, 
    selectedAnswers, 
    setSelectedAnswers
  } = useAppContext()
  
  const selectedAnswer = (answer) => answer.selectedAnswer !== answers
  const isSelected = selectedAnswers.every(selectedAnswer)

  const selectedAnswerClass = 
  !isPlayAgain ? 
  (isSelected ? "answer-button" : "selected-answer") : 
  (isSelected ? "answer-button" : "incorrect-answer")

  const buttonClass = 
  (isPlayAgain && (correctAnswer === answers)) ? 
  "correct-answer" : 
  selectedAnswerClass

  const decodedAnswers = Buffer.from(answers, 'base64').toString()
  function selectAnswer(){
    setSelectedAnswers(prevState => prevState.map(item => {
      return correctAnswer === item.id ? {...item, selectedAnswer: answers} : {...item}
    }))
  }

  return (
        <button 
          className={buttonClass} 
          onClick={selectAnswer} 
          disabled={isPlayAgain}
        >
          {decodedAnswers}
        </button>
  )
}