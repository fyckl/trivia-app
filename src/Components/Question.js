import React, { useState, useEffect, useContext } from 'react'
import Answer from './Answer'
import { Context, useAppContext } from "./Context"
import { Buffer } from 'buffer'
import useShuffleArr from '../Hooks/useShuffleArr'

function Question({question, incorrectAnswers, correctAnswer}) {
  
  const arrAnswers = incorrectAnswers.concat(correctAnswer)
  const [shuffledAnswers, setShuffledAnswers] = useState([])
  const { isShuffleAnswers } = useAppContext()
  const shuffleArr = useShuffleArr(arrAnswers)

  useEffect(() => {
    if(isShuffleAnswers){
      setShuffledAnswers(shuffleArr)
    }
  }, [isShuffleAnswers])

  const answers = shuffledAnswers.map((item) => {
      return <Answer key={item} answers={item} correctAnswer={correctAnswer}/>
  })
  const decodedQuestion = Buffer.from(question, 'base64').toString()
  return (
    <div className='question-container' data-testid="container">
        <h3 className='question'>{decodedQuestion}</h3>
        {answers}
        <hr className='hr'/>
    </div>
  )
}

export default Question