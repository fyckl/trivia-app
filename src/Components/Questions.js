import React, { useContext } from 'react'
import Question from './Question'
import useMappedQuestions from '../Hooks/useMappedQuestions'
import { Context, useAppContext } from "./Context"

export default function Questions(props) {
  const {
    arrQuestions, 
    checkAnswers, 
    isPlayAgain, 
    correctAnswers, 
    reset, 
    score
  } = useAppContext()

  // function questions(arr){ arr.map(item => (
  //   <Question 
  //     key={item.question}
  //     question={item.question} 
  //     correctAnswer={item.correct_answer} 
  //     incorrectAnswers={item.incorrect_answers}
  //   />
  // ))
  // }
  
  return (
    <div className='text-center questions-container' data-testid="container">
      {useMappedQuestions(arrQuestions)}

      {
        isPlayAgain ? 
        <div className='flex' id="post-game-details-container">
          <h4 className='inter'>You got {correctAnswers}/5 questions right</h4>
          <button className='again-button' onClick={reset}>Play again</button>
        </div> : 
        <button className='check-button' onClick={checkAnswers} id="check-answers">Check answers</button>
      }

      <h4 className='inter score'>Score: {score}</h4>
    </div>
  )
}
