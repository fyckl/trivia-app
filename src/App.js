import React, { useState, useEffect, useContext } from "react";
import Start from "./Components/Start";
import Questions from "./Components/Questions";
import Answers from "./Components/Answer";
import { Context } from "./Components/Context"
function App() {
  const {quizStart, startQuiz} = useContext(Context)

  return (
    <div className="main-container">
      {startQuiz ? <Questions /> : <Start start={quizStart}/>}
    </div>
  );

}

export default App;
