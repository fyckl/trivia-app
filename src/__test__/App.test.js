import React from "react"
import App from "../App"
import * as Context  from "../Components/Context"
import { fireEvent, cleanup, screen, render } from "@testing-library/react"
import "@testing-library/jest-dom/extend-expect"
import { Enzyme, shallow, configure, mount } from 'enzyme'
import Adapter from 'enzyme-adapter-react-16';

configure({ adapter: new Adapter() });

test("renders correctly to match snapshot", () => {
    const contextValues = {
        startQuiz: jest.fn(), 
        isStartQuiz: false, 
        arrQuestions:[], 
        isAPIError: false
    }

    jest
      .spyOn(Context, 'useAppContext')
      .mockImplementation(() => contextValues);

    const wrapper = shallow(<App/>)
    expect(wrapper.html()).toMatchSnapshot()
    expect(wrapper.find(App)).toBeTruthy();

   
})

test("renders correctly to match snapshot if there is an API error", () => {
    const contextValues = {
        startQuiz: jest.fn(), 
        isStartQuiz: false, 
        arrQuestions:[], 
        isAPIError: true
    }

    jest
      .spyOn(Context, 'useAppContext')
      .mockImplementation(() => contextValues);

    const wrapper = shallow(<App/>)
    expect(wrapper.html()).toMatchSnapshot()
    expect(wrapper.find(App)).toBeTruthy();

   
})

test("renders correctly to match snapshot if isStartQuiz is true", () => {
    const contextValues = {
        startQuiz: jest.fn(), 
        isStartQuiz: true, 
        arrQuestions:[], 
        isAPIError: true
    }

    jest
      .spyOn(Context, 'useAppContext')
      .mockImplementation(() => contextValues);

    const wrapper = shallow(<App/>)
    expect(wrapper.html()).toMatchSnapshot()
    expect(wrapper.find(App)).toBeTruthy();

   
})

test("renders correctly to match snapshot if arrQuestions length is 5 and isStartQuiz true", () => {
  const contextValues = {
      startQuiz: jest.fn(), 
      isStartQuiz: true, 
      arrQuestions:[1,2,3,4,5], 
      isAPIError: true
  }

  jest
    .spyOn(Context, 'useAppContext')
    .mockImplementation(() => contextValues);

  const wrapper = shallow(<App/>)
  expect(wrapper.html()).toMatchSnapshot()
  expect(wrapper.find(App)).toBeTruthy();

 
})