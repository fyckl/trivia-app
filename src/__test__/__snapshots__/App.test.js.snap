// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly to match snapshot 1`] = `"<div class=\\"main-container\\"><div class=\\"text-center questions-container\\" data-testid=\\"container\\"><h2 class=\\"karla\\" data-testid=\\"header\\">Quizzical</h2><h3 class=\\"karla\\" data-testid=\\"header-2\\">A game of trivia</h3><button name=\\"start-quiz\\" class=\\"start-button\\" data-testid=\\"button\\">Start Quiz</button></div></div>"`;

exports[`renders correctly to match snapshot if arrQuestions length is 5 and isStartQuiz true 1`] = `"<div class=\\"main-container\\"><h3 class=\\"karla text-center\\">Internal server error, please try again later (┬┬﹏┬┬)</h3></div>"`;

exports[`renders correctly to match snapshot if isStartQuiz is true 1`] = `"<div class=\\"main-container\\"><h3 class=\\"karla text-center\\">Internal server error, please try again later (┬┬﹏┬┬)</h3></div>"`;

exports[`renders correctly to match snapshot if there is an API error 1`] = `"<div class=\\"main-container\\"><h3 class=\\"karla text-center\\">Internal server error, please try again later (┬┬﹏┬┬)</h3></div>"`;
